docs/coding-conventions.md

```diff
@@ -0,0 +1,89 @@
+# Coding conventions
+
+This document outlines a collection of guidelines, style suggestions, and tips
+for writing code in the KubeVirt project. It is partially based on [Kubernetes
+Coding Conventions](https://github.com/kubernetes/community/blob/master/contributors/guide/coding-conventions.md).
+
+- [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
+- [Effective Go](https://golang.org/doc/effective_go.html)
+- Know and
+  avoid [Go landmines](https://gist.github.com/lavalamp/4bd23295a9f32706a48f)
+- Inline variable assignments when possible
+    - For example inline `err` checks
+- Use early returns to avoid unnecessarily deeply nested if/else blocks
+- Use switch-cases to avoid long if/else if/else blocks
+- Avoid named return values
+    - Define variables in the function body instead
+- Avoid closures when they do not capture any meaningful state
```

Comment:
Another useful use case of closures is to define a function that is relevant only for a very specific scope.
Since go doesn't have private variables and every function is shared with the whole package this is sometimes important IMO

Reply:
>The solution for a possible in-package conflict is to ensure that packages have a maintainable size (not too many files, not too long files) and that functions have clear useful names.

I agree.
However, sometimes even with small package sizes a helper function is relevant only for a specific scope.

I have two examples in mind for these cases.

The first: Let's say that a function needs to do an operation for a few times. Let's say that this operation is specific to the function's operation and has no meaning outside of it. In this scenario it makes total sense to wrap this helper operation within a variable-function and use it within the parent function when needed. The alternative of extracting the function is both error-prone (as someone might use it) and confusing (since someone might spend time on reasoning about this function in a scope where it doesn't make sense).

The second: tests. Usually there's more than one test in a package, and with unit tests it's almost a must since the test files reside in the production file-tree. It's very common that a very weird operation would be needed in order to test something, but that it wouldn't have any meaning outside the specific test's context.

-----

I guess that my problem is with the broad statement that **variable functions are a bad thing**. Yes, sometimes they are, but sometimes they do make sense. When making such a policy that would later be enforced in code reviews we need to be careful from such broad statements.

Reply:
I would suggest:
```suggestion
- Use closures with caution, be aware of the risks and use them only when it makes sense.
```

Reply:
The text of this page should state that maintainers are allowed to make well-motivated exceptions. But each convention should be a broad opinionated statement. I like @fabiand 's idea of adding examples for acceptable code and unacceptable one.  @iholder101 I'm certain you can provide a good use for variable functions. I can provide several problematic ones in https://github.com/kubevirt/kubevirt/blob/247ee51/tests/operator/operator.go#L125

Reply:
Challenge accepted!
I'll also try to fetch code that I have specifically wrote.

------

**For my first example above**

Here's a function that generates the error and being used multiple times in the parent function. It must be used only within this scope and has no meaning outside of it:
https://github.com/kubevirt/kubevirt/blob/aa5a345632ba9ccf7b282f107b049ed13d000514/pkg/virtctl/create/clone/clone.go#L261-L299

Another example:
https://github.com/kubevirt/kubevirt/blob/aa5a345632ba9ccf7b282f107b049ed13d000514/pkg/virt-handler/container-disk/mount.go#L554-L623

-----

**For my second example above**

There are many many examples, but here's one. The helper function is relevant for a specific test. Extracting it would make it harder to use and would bring no value to anyone outside the test:
https://github.com/kubevirt/kubevirt/blob/aa5a345632ba9ccf7b282f107b049ed13d000514/tests/migration/migration.go#L1212-L1230

-----

Another example is for documentation's sake. As I've explained in detail [here](https://github.com/kubevirt/kubevirt/issues/11446), I'm not a fan of comments. Sometimes it's better to define a one-time function just to give it a name and avoid using comments. An example from the above code is this:
```go
 		removeSliceElement := func(s []vmiMountTargetEntry, idxToRemove int) []vmiMountTargetEntry { 
 			s[idxToRemove] = s[len(s)-1] 
 			return s[:len(s)-1] 
 		} 
  
 		record.MountTargetEntries = removeSliceElement(record.MountTargetEntries, idx) 
 		return nil 
```

I think this is better than:
```go
		// removes slice element efficiently 
		record.MountTargetEntries[idx] = record.MountTargetEntries[len(s)-1] 
		record.MountTargetEntries = record.MountTargetEntries[:len(s)-1] 
 		return nil 
```

(note: there is an unnecessary comment in the actual code as well. It's a mistake IMO).

-----

Another valuable use is to use defers:
https://github.com/kubevirt/kubevirt/blob/aa5a345632ba9ccf7b282f107b049ed13d000514/cmd/virt-launcher-monitor/virt-launcher-monitor.go#L424-L450

Reply:
@iholder101 Agree, I did not want to express `variable functions are a bad thing` in general. But instead they should be used if they provide more value than a regular function. We have places in the code where you can simply move lambdas to regular functions without any changes necessary, and we should avoid such uses.